File: src/org/likexin/twopointer/HasCycle.java
package org.likexin.twopointer;

public class HasCycle {

  /**
   * 使用快慢指针，如果链表有环，那么slow和fast总会相遇的。
   * @param head
   * @return
   */
  public boolean hasCycle(ListNode head) {
    if (head == null || head.next == null) {
      return false;
    }

    ListNode slow = head, fast = head.next;
    while (fast.next != null && fast != null) {
      if (slow == fast) {
        return true;
      }
      slow = slow.next;
      fast = fast.next.next;
    }
    return false;
  }
}

File: Python/linkedlist/HasCycle.py
# -*- coding:UTF-8 -*-"""
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    """
    @ param: head: The first node of linked list.
    @ return: True if it has a cycle, or false
    @ solution:
    - 定义两个指针slow和fast，一个一次走一步，一个一次走两步，如果有环的话，一定会相遇；
    """
    def hasCycle(self, head):
        # write your code here
        if not head:
            return False 
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False
c